### PR TITLE
mock: in order mock calls

### LINKED ISSUE #1639: (syntax sugar) Implement an `InOrder` func for mocks
## Description

Since [gomock](https://github.com/golang/mock) has been archived, I am switching to testify's mock pkg.

One extremely neat feature that gomock had is the `InOrder` func:
https://pkg.go.dev/github.com/golang/mock/gomock#InOrder

Which allowed to very easily enforce expectation order by going from:
```go
someMock.EXPECT().ValidateArgs(gomock.Any(), 1).Return(nil),
someMock.EXPECT().ValidateArgs(gomock.Any(), 2).Return(nil)
```
to
```go
gomock.InOrder(
    someMock.EXPECT().ValidateArgs(gomock.Any(), 1).Return(nil),
    someMock.EXPECT().ValidateArgs(gomock.Any(), 2).Return(nil)
)
```

As opposed to the current API implemented in testify which is more cumbersome and less intuitive with the need to declare vars and pass them inside the last chained method.
https://github.com/stretchr/testify/issues/741
https://github.com/stretchr/testify/pull/1106

```go
call1 := mockThing.On("Init").Return(nil)
call2 := mockThing.On("Do").Return(nil).NotBefore(call1)
mockThing.On("Close").Return(nil).NotBefore(call1, call2)
```

Or I guess do some kind of nesting which reverse the chronological order of events and creates nesting.
```go
mockThing.On("Close").Return(nil).NotBefore(
    mockThing.On("Do").Return(nil).NotBefore(
        mockThing.On("Init").Return(nil).NotBefore(
            ... nesting
)
```

Ideally could be turned into:

```go
mock.InOrder(
    mockThing.On("Init").Return(nil),
    mockThing.On("Do").Return(nil),
    mockThing.On("Close").Return(nil)
)
```

## Proposed solution

I didn't looked into the codebase yet to know if that's easily feasible. 
But my gut feeling says it could be trivial to chain the `NotBefore` inside the `mock.InOrder` variadic function.
This issue is mostly to gather feedback.

## Use case

More intuitive API.

### PR DESCRIPTION
**Pull request by ReyOrtiz**
## Summary

Add helper method to declare the order of the mock calls.

## Changes

* add `InOrder(calls ...*Call)` helper method to declare the order of mock calls by calling `NotBefore()`

## Motivation

This is probably the only feature I miss from `gomock.InOrder()`, it is more intuitive and cleaner to declare the order of the mock calls. Probably most cases can be satisfied with this method.

```
InOrder(
	mockedService1.On("TheExampleMethod", 1, 2, 3).Return(4, nil),
	mockedService2.On("TheExampleMethod2", true).Return(),
	mockedService3.On("TheExampleMethod3", "abc", 3).Return(false, nil),
)

```
## Related issues
Closes #1639

### CODE REVIEW

**Comment by brackendawson** [comment]
Ah, an earlier issue, I suspect we'll see this a lot from gomock expats over the coming months.

I'm going to close this is a duplicate of #1639. I know this one was first but the other is more in depth. Please do go add your 👍 to it.

**Review by brackendawson — CHANGES_REQUESTED** [review]
Can we get a unit test, please.

**Comment by ReyOrtiz** [comment]
unit test added

File: mock/mock_test.go
Comment (brackendawson) [inline mock/mock_test.go:955]: This line is testing the implementation. A cost effective test would only test the behaviour, as you are doing just fine after this line.

If you remove this assertion then we are free to change the implementation of the intended behaviour without breaking the test. It also removes the need to declare `b` and `c`, so the test can be written as `InOrder` is intended to be used.

File: mock/mock.go
Comment (brackendawson) [inline mock/mock.go:276]: This could do with a small usage example for the docs.

**Review by brackendawson — CHANGES_REQUESTED** [review]
Positive and negative tests, nice one. 👍  Just a couple of small comments.

File: mock/mock_test.go
Comment (brackendawson) [reply mock/mock_test.go:955]: Also, it's just comparing pointer values. If InOrder did literally nothing then that assertion would still pass.

File: mock/mock.go
Comment (brackendawson) [reply mock/mock.go:276]: It's also the order they need to be "satisfied", not the order they need to be called. In the case where they can be called multiple times.

**Comment by ReyOrtiz** [comment]
Good comments, thanks! Requested changes applied.

**Review by brackendawson — APPROVED** [review]
Looks good to me, should be helpful to all the developers moving from gomock. I'll give some time for other maintainers to review before I merge.

**Comment by jybp** [comment]
Ah you beat me to it, thanks!

**Comment by ReyOrtiz** [comment]
Hello, any ETA on merging this PR?

**Comment by brackendawson** [comment]
Thanks @ReyOrtiz. This can be consumed now by gomock expats with: `go get github.com/stretchr/testify@master`. Once testify v1.10 is released it can be consumed as per the README. I'll have a look at drafting a release soon.

**Comment by ReyOrtiz** [comment]
Thanks @brackendawson for all the help! 🙏

**Pull request merged**
